File: modules/common/ipc/examples/py/pub_example.py
# =================================================================================================
# Copyright (C) 2025 GRAPE Contributors
# =================================================================================================

import time
from grape_ipc_py import init, Config, RawPublisher, ok


def to_bytes(msg: str) -> bytes:
    """Convert a string to bytes."""
    return msg.encode("utf-8")


def match_callback(match):
    """Callback for match/unmatch events."""
    print(f"Match status: {match.status}")


def main():
    try:
        # Initialize the IPC session
        config = Config()
        config.name = "Python IPC publisher example"
        init(config)

        # Define the topic
        topic = "hello_world"

        # Create the publisher
        publisher = RawPublisher(topic, match_callback)

        # Publish messages in a loop
        counter = 0
        SLEEP_TIME = 0.5  # 500 milliseconds
        while ok():
            counter += 1
            message = f"Hello World {counter}"
            print(f"Sending message: '{message}'")
            publisher.publish(to_bytes(message))
            time.sleep(SLEEP_TIME)

    except Exception as e:
        print(f"An error occurred: {e}")


if __name__ == "__main__":
    main()

File: modules/common/ipc/examples/pub_example.cpp
//=================================================================================================
// Copyright (C) 2025 GRAPE Contributors
//=================================================================================================

#include <cstddef>
#include <print>
#include <thread>

#include "grape/exception.h"
#include "grape/ipc/publisher.h"
#include "grape/ipc/session.h"
#include "topic_example.h"

//=================================================================================================
// Demonstrates a basic IPC publisher. See sub_example.cpp for the corresponding subscriber.
auto main() -> int {
  try {
    grape::ipc::init(grape::ipc::Config{});

    const auto match_cb = [](const grape::ipc::Match& match) -> void {
      std::println("\n{} (entity: {})", toString(match.status), toString(match.remote_entity));
    };

    const auto topic = grape::ipc::ex::ExampleTopicAttributes{};
    auto publisher = grape::ipc::Publisher(topic, match_cb);

    auto counter = 0U;
    constexpr auto SLEEP_TIME = std::chrono::milliseconds(500);
    while (grape::ipc::ok()) {
      const auto message = std::string("Hello World ") + std::to_string(++counter);
      std::println("Sending message: '{}'", message);
      const auto status = publisher.publish(message);
      if (not status) {
        std::println("Publish failed: {}", toString(status.error()));
      }
      std::this_thread::sleep_for(SLEEP_TIME);
    }
    return EXIT_SUCCESS;
  } catch (...) {
    grape::Exception::print();
    return EXIT_FAILURE;
  }
}
